mlflow/store/azure_blob_artifact_repo.py

```diff
@@ -0,0 +1,123 @@
+    """
+    Stores artifacts on Azure Blob Storage.
+
+    This repository is used with URIs of the form
+    ``wasbs://<container-name>@<ystorage-account-name>.blob.core.windows.net/<path>``,
+    following the same URI scheme as Hadoop on Azure blob storage. It requires that your Azure
+    storage access key be available in the environment variable ``AZURE_STORAGE_ACCESS_KEY``.
+    """
+
+    def __init__(self, artifact_uri, client=None):
+        if client:
+            self.client = client
+        else:
+            from azure.storage.blob import BlockBlobService
+            (_, account, _) = AzureBlobArtifactRepository.parse_wasbs_uri(artifact_uri)
+            if "AZURE_STORAGE_CONNECTION_STRING" in os.environ:
+                self.client = BlockBlobService(
+                    account_name=account,
+                    connection_string=os.environ.get("AZURE_STORAGE_CONNECTION_STRING"))
+            elif "AZURE_STORAGE_ACCESS_KEY" in os.environ:
+                self.client = BlockBlobService(
+                    account_name=account,
+                    account_key=os.environ.get("AZURE_STORAGE_ACCESS_KEY"))
+            else:
+                raise Exception("You need to set one of AZURE_STORAGE_CONNECTION_STRING or "
+                                "AZURE_STORAGE_ACCESS_KEY to access Azure storage.")
+        super(AzureBlobArtifactRepository, self).__init__(artifact_uri)
+
+    @staticmethod
+    def parse_wasbs_uri(uri):
+        """Parse a wasbs:// URI, returning (container, storage_account, path)."""
+        parsed = urllib.parse.urlparse(uri)
+        if parsed.scheme != "wasbs":
+            raise Exception("Not a WASBS URI: %s" % uri)
+        match = re.match("([^@]+)@([^.]+)\\.blob\\.core\\.windows\\.net", parsed.netloc)
+        if match is None:
+            raise Exception("WASBS URI must be of the form "
+                            "<container>@<account>.blob.core.windows.net")
+        container = match.group(1)
+        storage_account = match.group(2)
+        path = parsed.path
+        if path.startswith('/'):
+            path = path[1:]
+        return container, storage_account, path
+
+    def log_artifact(self, local_file, artifact_path=None):
+        (container, _, dest_path) = self.parse_wasbs_uri(self.artifact_uri)
+        if artifact_path:
+            dest_path = build_path(dest_path, artifact_path)
+        dest_path = build_path(dest_path, os.path.basename(local_file))
+        self.client.create_blob_from_path(container, dest_path, local_file)
+
+    def log_artifacts(self, local_dir, artifact_path=None):
+        (container, _, dest_path) = self.parse_wasbs_uri(self.artifact_uri)
+        if artifact_path:
+            dest_path = build_path(dest_path, artifact_path)
+        local_dir = os.path.abspath(local_dir)
+        for (root, _, filenames) in os.walk(local_dir):
+            upload_path = dest_path
+            if root != local_dir:
```

Comment:
This seems a bit complicated -- is there a unit test covering the case where `root == local_dir` and where `root != local_dir`? I guess this should be included as long as the unit test includes a subdirectory.